File: 剑指Offer33-er-cha-sou-suo-shu-de-hou-xu-bian-li-xu-lie-lcof/33.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from typing import List

class Solution:
    def verifyPostorder(self, postorder: List[int]) -> bool:
        def recr(l: int, r: int) -> bool:
            if l >= r: return True
            i = l
            while postorder[i] < postorder[r]: i += 1
            mid = i
            while i < r and postorder[i] > postorder[r]: i += 1
            return i == r and recr(l, mid-1) and recr(mid, r-1)
        
        return recr(0, len(postorder)-1)


if __name__ == '__main__':
    sol = Solution()
    print(sol.verifyPostorder([1,6,3,2,5]))
    print(sol.verifyPostorder([1,3,2,6,5]))

File: shoope/剑指Offer33-er-cha-sou-suo-shu-de-hou-xu-bian-li-xu-lie-lcof/33.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    bool verifyPostorder(vector<int>& postorder) {
        return recv(postorder, 0, postorder.size()-1);
    }

    bool recv(vector<int>& postorder, int l, int r) {
        if (l >= r) return true;
        int p = l;
        while (postorder[p] < postorder[r]) ++p;
        int mid = p;
        while (postorder[mid] > postorder[r]) ++mid;
        return mid == r && recv(postorder, l, p-1) && recv(postorder, p, r-1);
    }
};

int main(int argc, char** args) {
    Solution s = Solution();
    vector<int> arr = {1,6,3,2,5};
    cout << s.verifyPostorder(arr) << endl;
    arr = {1,3,2,6,5};
    cout << s.verifyPostorder(arr) << endl;
    return 0;
}